File: code/algos/gohelper.py
"""
Author : Rupesh Kumar
Date   : June 11th 2020
File   : tojangoPlane.py

Description : Define Player and Point class.

"""

import enum
from collections import namedtuple
import operator

__all__ = [
    'Player',
    'Point',
    'is_point_an_eye',
]

def directionVector_add(a, b):
    return tuple(map(operator.add, a, b))

directions = {'W': (-1, 0), 'E': (1, 0), 'S': (0, -1), 'N': (0, 1)}
action = {1 : 'W', 2 : 'E', 3 : 'S', 4 : 'N'}


class Player(enum.Enum):
    black = 1
    white = 2

    @property
    def opp(self):  #opposition
        if self == Player.white:
            return Player.black 
        else:
            return Player.white

class Point(namedtuple('Point', 'row col')):
    def neighbors(self):
        neigh = []
        for i in range (1, len(directions) + 1):
            row, col = directionVector_add((self.row, self.col),(directions[action[i]]))
            neigh.append(Point(row, col))
            
        return neigh

    def __deepcopy__(self, memodict={}):
        return self



def is_point_an_eye(board, point, player):
    """
    Checks if the point is an eye for the player being passed in (not evaluating if point is an eye for either player). 
    Checks if point is:
        1) a liberty
        2) surrounded by a chain or almost chain
            a) neighbors are the same color
            b) neighbors are part of the same chain
    """
    r, c = point
    if board.grid[r][c] != 0:
        return False

    neighs = point.neighbors()

    chains = set()
    visited = set()
    adj_libs = set()
            
    # neighbors are the same color and are part of a chain
    for n in neighs: 
        nr, nc = n

        # if neighbor is invalid then we skip computations
        if nr < 0 or nc < 0 or not board.is_on_grid(Point(row=nr, col=nc)):
            continue

        # if a neighbor isn't occupied by current player, then it's not an eye
        if board.grid[nr][nc] != player.value:
            return False

        if n not in visited:
            curr_chain, adj_lib = find_connected(board, n, player)
            visited = visited.union(curr_chain)
            if len(adj_libs) > 0:
                adj_libs = adj_libs.intersection(adj_lib)
            else:
                adj_libs = adj_lib
            chains.add(tuple(curr_chain))

    if len(chains) == 1:
        return True

    # print(chains)
    # print(adj_libs)
    lib_count = 0
    for lib in adj_libs:
        if lib == point:
            continue
        # print(lib)
        board.grid[lib[0]][lib[1]] = player.value
        c, _ = find_connected(board, lib, player)
        ss_count = 0
        for chain in chains:
            # print("c %s" % c)
            # print("chain %s" % str(chain))
            # print(set(chain).issubset(c))
            if set(chain).issubset(c):
                ss_count += 1
            else:
                break
        # print(ss_count)
        if ss_count == len(chains):
            # print(lib)
            lib_count += 1
        board.grid[lib[0]][lib[1]] = 0
    # print(lib_count)
    if lib_count > 1:
        return True
    return False
    
def find_connected(board, point, player):
    """
    BFS from starting point to get all pieces in a connected group. 
    """
    #print("point, player : ", point, player)
    queue = [point]

    visited = set() # you're only visiting neighbors in same chain
    adj_lib = set()
    while len(queue) > 0:
        curr = queue.pop(0)
        neighs = curr.neighbors()
        visited.add(curr)
        for n in neighs:
            r, c = n
            if not board.is_on_grid(point) or not board.is_on_grid(Point(row=r, col=c)):
                continue
            if n not in visited and board.grid[r][c] == player.value:
                queue.append(n)
            if board.grid[r][c] == 0:
                adj_lib.add(n)
    return visited, adj_lib

"""
#Driver code
if __name__ == "__main__":

    #usage of Player class. Output is mentioned after # for each print statement, so use accordingly.
    print("Player : ", Player.black, Player.black.value) #Player :  Player.black 1
    print("Player : ", Player.white, Player.white.value) #Player :  Player.white 2


    self_player = Player.black
    print(self_player, self_player.value) #Player.black 1

    opp_player = self_player.opp
    print(opp_player, opp_player.value) #Player.white 2
    

    #usage of Point
    print("Point is : ", Point(row=1, col=1)) # Point is :  Point(row=1, col=1)

    row = 2
    col = 2
    print("Point is : ", Point(row, col)) # Point is :  Point(row=2, col=2)

    point = Point(row=3, col=3)
    print(point) # Point(row=3, col=3)

    #Find the neighbors of Point (row=3,col=3). Any point have 4 neighbors
    neighs = point.neighbors()
    for neigh in neighs:
        print(neigh) # Point(row=2, col=3) Point(row=4, col=3) Point(row=3, col=2) Point(row=3, col=4)
    
"""

File: code/algos/gohelper.cpp
/*
Author :  
Date   : July 30th 2020
File   : gohelper.cpp

Description : Define Go Domain Rules related helper functions and data-types.
*/

#include <iostream>
#include <list>
#include <set>
#include <vector>
#include <map>
#include <iterator>
#include <algorithm>
#include <functional>
#include <tuple>

#include "gohelper.h"

using namespace std;

// Default constructor
Point::Point()
{
    coord.first = -1;
    coord.second = -1;
}

// Parametrized constructor
Point::Point(int row, int col)
{
    coord.first = row;
    coord.second = col;
}

//object comparsion (==) operator overloading
bool Point::operator==(const Point &other) const
{
    cout << "Point == operator is called" << endl;
    return (coord.first == other.coord.first && coord.second == other.coord.second);
}

//object comparsion (<) operator overloading
bool Point::operator<(const Point &other) const
{
    cout << "Point < operator is called" << endl;
    return (coord.first < other.coord.first);
}

//Assignment(=) operator overloading
Point *Point::operator=(const Point &other)
{
    cout << "Point = operator is called" << endl;
    coord.first = other.coord.first;
    coord.second = other.coord.second;
    return this;
}

// Find the neighbours of the this point.
vector<Point> Point::neighbours()
{
    map<int, char> action = mapAction();
    map<char, Point> directions;
    vector<Point> neigh;
    int row, col;

    directions = mapDirections();
    size_t len = directions.size();
    for (int i = 1; i <= len; i++)
    {
        neigh.push_back(Point(coord.first + directions[action[i]].coord.first, coord.second + directions[action[i]].coord.second));
        //.first, coord.second + directions[action[i]].second));
        //cout << point.first + directions[action[i]].first << " " << point.second + directions[action[i]].second << endl;
    }

    for (auto &elm : neigh)
    {
        cout << elm.coord.first << " " << elm.coord.second << endl;
    }

    return neigh;
}

#ifdef USAGE
int main()
{

    vector<Point> neigh;

    neigh = Point(1, 1).neighbours();

    cout << "Print the neighbours here ..." << endl;
    for (auto &elm : neigh)
    {
        cout << elm.coord.first << " " << elm.coord.second << endl;
    }

    Point p1(1, 1);
    Point p2(1, 1);

    if (p1 == p2)
    {
        cout << "both points are equal" << endl;
    }
    else
    {
        cout << "both points are not equal" << endl;
    }

    Player player(BLACK);
    cout << "OPPOSITION : " << player.opp() << endl;

    return 0;
}
#endif
